### ISSUE TITLE
file.managed gives UnboundLocalError: local variable 'ret' referenced before assignment

### ISSUE THREAD

**Issue by anlutro** [saltstack/salt#47344]
I'm running off a git checkout of the 2018.3 branch.

I have this state:

```
/etc/motd:
  file.managed:
    - contents: {{ pillar.get('motd', '') | json }}
```

```
          ID: /etc/motd
    Function: file.managed
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/var/tmp/.salt_08831c_salt/py3/salt/state.py", line 1905, in call
                  **cdata['kwargs'])
                File "/var/tmp/.salt_08831c_salt/py3/salt/loader.py", line 1830, in wrapper
                  return f(*args, **kwargs)
                File "/var/tmp/.salt_08831c_salt/py3/salt/states/file.py", line 2396, in managed
                  **kwargs
                File "/var/tmp/.salt_08831c_salt/py3/salt/modules/file.py", line 4738, in check_managed_changes
                  group, mode, attrs, saltenv, contents)
                File "/var/tmp/.salt_08831c_salt/py3/salt/modules/file.py", line 4845, in check_file_meta
                  differences = get_diff(name, tmp, show_filenames=False)
                File "/var/tmp/.salt_08831c_salt/py3/salt/modules/file.py", line 5011, in get_diff
                  return ret
              UnboundLocalError: local variable 'ret' referenced before assignment
     Started: 01:13:42.231258
    Duration: 39.385 ms
     Changes:  
```

Looking briefly at the code in question, it looks like the `ret` variable is only set inside of an if statement which might not always be true, but I wasn't able to figure out what the purpose of it was, so I couldn't submit a PR.

```
$ salt-ssh --versions
Salt Version:
           Salt: 2018.3.0-362-g245d62c
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.5.3 (default, Jan 19 2017, 14:11:04)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 17.0.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.1.6
 
System Versions:
           dist: debian 9.4 
         locale: UTF-8
        machine: x86_64
        release: 4.9.0-6-amd64
         system: Linux
        version: debian 9.4
```

**Comment by The-Loeki** [comment]
https://github.com/saltstack/salt/blob/2017.7/salt/modules/file.py#L4656
https://github.com/saltstack/salt/blob/2018.3/salt/modules/file.py#L4994

From the looks of it `ret` should be `''` if there is no diff, which isn't happening due to the `args[0] == args[1]` being `True` when there are no differences (`args` gets filled in the loop above w/`file1` & `file2`).

**Comment by gtmanfred** [comment]
Thanks for the fix @The-Loeki 

Daniel

**Comment by terminalmage** [comment]
Sorry @The-Loeki, I never saw your PR, I received a comment on the PR I opened last week which caused this issue in the head of the 2018.3 branch, and made a separate fix (which includes a unit test): https://github.com/saltstack/salt/pull/47405

**Comment by rallytime** [comment]
Closed via #47405
